File: src/main/java/eladkay/quaritum/common/core/PositionedBlockHelper.java
package eladkay.quaritum.common.core;

import eladkay.quaritum.api.lib.LibNBT;
import eladkay.quaritum.api.rituals.PositionedBlock;
import eladkay.quaritum.api.util.ItemNBTHelper;
import eladkay.quaritum.common.block.ModBlocks;
import eladkay.quaritum.common.item.ModItems;
import net.minecraft.block.properties.IProperty;
import net.minecraft.block.state.IBlockState;
import net.minecraft.item.EnumDyeColor;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3i;
import net.minecraft.world.World;

import java.util.List;

public final class PositionedBlockHelper {

    public static PositionedBlock positionedBlockWith(BlockPos pos, EnumDyeColor chalkColor) {
        return PositionedBlock.constructPosChalk(chalkColor, pos);
    }

    public static int getChalkPriority(List<PositionedBlock> list, TileEntity entity, String optionalName) {
        World world = entity.getWorld();
        int chalks = 0;
        for (PositionedBlock block : list) {
            IBlockState state = world.getBlockState(entity.getPos().add(block.getPos()));
            List<IProperty> comparables = block.getComparables();

            if (block.getState().getBlock() == state.getBlock()) {
                LogHelper.logDebug("Block check OK for " + optionalName);
                if ((block.getState().getBlock() == ModBlocks.tempest && state.getBlock() == ModBlocks.tempest))
                    chalks++;
                else if (comparables != null) {
                    for (IProperty property : comparables) {
                        if (block.getState().getValue(property) != state.getValue(property)) {
                            LogHelper.logDebug("Expected " + block.getState() + " in " + block.getPos() + " for ritual " + optionalName + ". Got " + state);
                            return -1;
                        }
                    }
                    chalks++;
                } else if (block.getState() != state) {
                    LogHelper.logDebug("Expected " + block.getState() + " in " + block.getPos() + " for ritual " + optionalName + ". Got " + state);
                    return -1;
                } else chalks++;
            } else {
                return -1;
            }
        }

        return chalks;
    }

    public static Vec3i getDimensions(List<PositionedBlock> list) {
        int minX = 0;
        int maxX = 0;
        int minY = 0;
        int maxY = 0;
        int minZ = 0;
        int maxZ = 0;
        for (PositionedBlock block : list) {
            minX = Math.min(block.getPos().getX(), minX);
            maxX = Math.max(block.getPos().getX(), maxX);

            minY = Math.min(block.getPos().getY(), minY);
            maxY = Math.max(block.getPos().getY(), maxY);

            minZ = Math.min(block.getPos().getZ(), minZ);
            maxZ = Math.max(block.getPos().getZ(), maxZ);
        }
        return new Vec3i(maxX - minX, maxY - minY, maxZ - minZ);
    }

    @Deprecated
    public static ItemStack[][][] getItemStackArrayArrayArrayFromPositionedBlockList(List<PositionedBlock> blocks) {
        ItemStack[][][] ret = new ItemStack[50][50][50];
        for (PositionedBlock block : blocks) {
            ItemStack stack = new ItemStack(ModItems.chalk, 1, block.getState().getBlock().getMetaFromState(block.getState()));
            ret[block.getPos().getX() + 3][block.getPos().getY() + 3][block.getPos().getZ() + 3] = stack;
        }
        return ret;
    }

    public static ItemStack getStackFromChalk(PositionedBlock block) {
        if (block.getState().getBlock() == ModBlocks.blueprint) return new ItemStack(ModItems.picture, 1, 0);
        return new ItemStack(ModItems.chalk, 1, block.getState().getBlock().getMetaFromState(block.getState()));
    }

    public static ItemStack getStackFromChalk(PositionedBlock block, boolean flat) {
        ItemStack ret = new ItemStack(block.getState().getBlock(), 1, block.getState().getBlock().getMetaFromState(block.getState()));
        if (block.getState().getBlock() == ModBlocks.chalk)
            ret = new ItemStack(ModItems.chalk, 1, block.getState().getBlock().getMetaFromState(block.getState()));
        else if (block.getState().getBlock() == ModBlocks.tempest) ret = new ItemStack(ModItems.tempest, 1);
        if (flat)
            ItemNBTHelper.setBoolean(ret, LibNBT.FLAT, true);

        return ret;
    }

}

File: src/main/java/eladkay/quaeritum/common/core/PositionedBlockHelper.kt
package eladkay.quaeritum.common.core

import eladkay.quaeritum.api.rituals.PositionedBlock
import eladkay.quaeritum.common.block.ModBlocks
import net.minecraft.block.properties.IProperty
import net.minecraft.block.state.IBlockState
import net.minecraft.item.EnumDyeColor
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3i

object PositionedBlockHelper {

    fun positionedBlockWith(pos: BlockPos, chalkColor: EnumDyeColor): PositionedBlock {
        return PositionedBlock.constructPosChalk(chalkColor, pos)
    }

    fun transform(list: MutableList<PositionedBlock>, facing: EnumFacing, mirrorAlongX: Int) {
        if (facing == EnumFacing.NORTH && mirrorAlongX == 1) return
        for (i in list.indices)
            list[i] = list[i].transform(facing, mirrorAlongX)
    }

    fun getChalkPriority(list: List<PositionedBlock>, entity: TileEntity): Int {
        val world = entity.world
        var chalks = 0
        for (block in list) {
            val state = world.getBlockState(entity.pos.add(block.pos))
            val comparables = block.comparables

            if (block.state.block == state.block) {
                when {
                    block.state.block == ModBlocks.tempest -> chalks++
                    comparables != null -> {
                        if (comparables.any { !castCheckValue(it, block.state, state) })
                            return -1
                        chalks++
                    }
                    block.state !== state -> return -1
                    else -> chalks++
                }
            } else {
                return -1
            }
        }

        return chalks
    }

    @Suppress("UNCHECKED_CAST")
    private fun castCheckValue(property: IProperty<*>, stateOne: IBlockState, stateTwo: IBlockState): Boolean {
        return checkValue(property as IProperty<Comparable<Comparable<*>>>, stateOne, stateTwo)
    }

    private fun <T : Comparable<T>> checkValue(property: IProperty<T>, stateOne: IBlockState, stateTwo: IBlockState): Boolean {
        return stateOne.getValue(property) == stateTwo.getValue(property)
    }

    fun getDimensions(list: List<PositionedBlock>): Vec3i {
        var minX = 0
        var maxX = 0
        var minY = 0
        var maxY = 0
        var minZ = 0
        var maxZ = 0
        for (block in list) {
            minX = Math.min(block.pos.x, minX)
            maxX = Math.max(block.pos.x, maxX)

            minY = Math.min(block.pos.y, minY)
            maxY = Math.max(block.pos.y, maxY)

            minZ = Math.min(block.pos.z, minZ)
            maxZ = Math.max(block.pos.z, maxZ)
        }
        return Vec3i(maxX - minX, maxY - minY, maxZ - minZ)
    }

}
